### ISSUE TITLE
Non-Deterministic Outputs of torch.compile's Model When using torch.flatten and torch.cat

### ISSUE THREAD

**Issue by ChijinZ** [pytorch/pytorch#109030]
### 🐛 Describe the bug

Let's consider this code snippet:

```python
import numpy as np
import torch

class M(torch.nn.Module):
    def __init__(self):
        super().__init__()
        self.v9_0 = torch.nn.Parameter(torch.tensor([4], dtype=torch.int8), requires_grad=False)
        self.v8_0 = torch.nn.Parameter(torch.tensor([5], dtype=torch.int8), requires_grad=False)
        self.v6_0 = torch.nn.Parameter(torch.tensor([6], dtype=torch.int8), requires_grad=False)


    def forward(self, *args):
        _args = args
        getitem = _args[0]
        getitem_1 = _args[1]
        getitem_2 = _args[2];  _args = None
        flatten = getitem.flatten();  getitem = None
        v6_0 = self.v6_0
        v8_0 = self.v8_0
        v9_0 = self.v9_0
        cat = torch.cat((getitem_1, v6_0, flatten, v8_0, v9_0), dim = 0);  getitem_1 = v6_0 = flatten = v8_0 = v9_0 = None
        sum_1 = cat.sum(0)
        sub = torch.sub(cat, getitem_2);  cat = getitem_2 = None
        return (sum_1, sub)

m = M()

inp = [np.array([[[[4]]]], dtype='int8'), np.array([3,6,5,7,3,6,3,4,7,5,6,4,4,4,4,5,7,4,7,4,7,6,5,7,6,6,3,4,6,7,5,5,5,4], dtype='int8'), np.array(7, dtype='int8')]

opt = torch.compile(m, fullgraph=True, backend='inductor', mode=None)

m_out = m(*[torch.from_numpy(v).to('cpu') for v in inp])
print(f"torch result: {m_out}")
opt_out = opt(*[torch.from_numpy(v).to('cpu') for v in inp])
print(f"torch.compile result: {opt_out}")
```

I run this script for several times, but the results are weird. The outputs of the model before ``torch.compile`` are deterministic, which are 
```
torch result: (tensor(193), tensor([-4, -1, -2,  0, -4, -1, -4, -3,  0, -2, -1, -3, -3, -3, -3, -2,  0, -3,
         0, -3,  0, -1, -2,  0, -1, -1, -4, -3, -1,  0, -2, -2, -2, -3, -1, -3,
        -2, -3], dtype=torch.int8))
```
The results are consistent across several runs. However, the outputs of the model after ``torch.compile`` are non-deterministic. Each time outputs a totally different result like
```
torch.compile result: (tensor(56), tensor([ 105, -106,   90,   19,    1,  120,   -7,   -7, -119, -100,  -84,  -38,
          73,   79,   -7,   -7,   25,   -7,   -7,   -7,   -7,   -7,   -7,   -7,
         121,   -7,   -7,   -7,   -7,   -7,   -7,   -7,   -2,   -3,   -1,   -3,
          -2,   -3], dtype=torch.int8))
```
```
torch.compile result: (tensor(-407), tensor([  20,   -3,  -11, -115,  -81,   83,  -55,  -33,   73, -112,  -62,  -50,
         -52,  120,   -7,   -7,   73,  -28,  -55,  -16,   51,   79,   -7,   -7,
          10,  -51,   36, -122,  123,   34, -128, -103,   -2,   -3,   -1,   -3,
          -2,   -3], dtype=torch.int8))
```

It seems that it triggers a undefined behavior in the inductor compiler.

### Versions

Collecting environment information...
PyTorch version: 2.0.1+cu117
Is debug build: False
CUDA used to build PyTorch: 11.7
ROCM used to build PyTorch: N/A

OS: Ubuntu 22.04.3 LTS (x86_64)
GCC version: (Ubuntu 11.4.0-1ubuntu1~22.04) 11.4.0
Clang version: 13.0.1-2ubuntu2.2
CMake version: version 3.27.2
Libc version: glibc-2.35

Python version: 3.10.12 (main, Jun 11 2023, 05:26:28) [GCC 11.4.0] (64-bit runtime)
Python platform: Linux-5.19.0-46-generic-x86_64-with-glibc2.35
Is CUDA available: True
CUDA runtime version: 12.1.105
CUDA_MODULE_LOADING set to: LAZY
GPU models and configuration: GPU 0: NVIDIA GeForce GTX 1660 SUPER
Nvidia driver version: 525.125.06
cuDNN version: Probably one of the following:
/usr/lib/x86_64-linux-gnu/libcudnn.so.8.9.1
/usr/lib/x86_64-linux-gnu/libcudnn_adv_infer.so.8.9.1
/usr/lib/x86_64-linux-gnu/libcudnn_adv_train.so.8.9.1
/usr/lib/x86_64-linux-gnu/libcudnn_cnn_infer.so.8.9.1
/usr/lib/x86_64-linux-gnu/libcudnn_cnn_train.so.8.9.1
/usr/lib/x86_64-linux-gnu/libcudnn_ops_infer.so.8.9.1
/usr/lib/x86_64-linux-gnu/libcudnn_ops_train.so.8.9.1
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

CPU:
Architecture:                    x86_64
CPU op-mode(s):                  32-bit, 64-bit
Address sizes:                   39 bits physical, 48 bits virtual
Byte Order:                      Little Endian
CPU(s):                          20
On-line CPU(s) list:             0-19
Vendor ID:                       GenuineIntel
Model name:                      12th Gen Intel(R) Core(TM) i7-12700KF
CPU family:                      6
Model:                           151
Thread(s) per core:              2
Core(s) per socket:              12
Socket(s):                       1
Stepping:                        2
CPU max MHz:                     5000.0000
CPU min MHz:                     800.0000
BogoMIPS:                        7219.20
Flags:                           fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush dts acpi mmx fxsr sse sse2 ss ht tm pbe syscall nx pdpe1gb rdtscp lm constant_tsc art arch_perfmon pebs bts rep_good nopl xtopology nonstop_tsc cpuid aperfmperf tsc_known_freq pni pclmulqdq dtes64 monitor ds_cpl vmx est tm2 ssse3 sdbg fma cx16 xtpr pdcm sse4_1 sse4_2 x2apic movbe popcnt tsc_deadline_timer aes xsave avx f16c rdrand lahf_lm abm 3dnowprefetch cpuid_fault epb ssbd ibrs ibpb stibp ibrs_enhanced tpr_shadow vnmi flexpriority ept vpid ept_ad fsgsbase tsc_adjust bmi1 avx2 smep bmi2 erms invpcid rdseed adx smap clflushopt clwb intel_pt sha_ni xsaveopt xsavec xgetbv1 xsaves split_lock_detect avx_vnni dtherm ida arat pln pts hwp hwp_notify hwp_act_window hwp_epp hwp_pkg_req hfi umip pku ospke waitpkg gfni vaes vpclmulqdq rdpid movdiri movdir64b fsrm md_clear serialize arch_lbr ibt flush_l1d arch_capabilities
Virtualization:                  VT-x
L1d cache:                       512 KiB (12 instances)
L1i cache:                       512 KiB (12 instances)
L2 cache:                        12 MiB (9 instances)
L3 cache:                        25 MiB (1 instance)
NUMA node(s):                    1
NUMA node0 CPU(s):               0-19
Vulnerability Itlb multihit:     Not affected
Vulnerability L1tf:              Not affected
Vulnerability Mds:               Not affected
Vulnerability Meltdown:          Not affected
Vulnerability Mmio stale data:   Not affected
Vulnerability Retbleed:          Not affected
Vulnerability Spec store bypass: Mitigation; Speculative Store Bypass disabled via prctl
Vulnerability Spectre v1:        Mitigation; usercopy/swapgs barriers and __user pointer sanitization
Vulnerability Spectre v2:        Mitigation; Enhanced IBRS, IBPB conditional, RSB filling, PBRSB-eIBRS SW sequence
Vulnerability Srbds:             Not affected
Vulnerability Tsx async abort:   Not affected

Versions of relevant libraries:
[pip3] numpy==1.24.3
[pip3] onnx2torch==1.5.11
[pip3] torch==2.0.1
[pip3] torchaudio==2.0.2
[pip3] torchsummary==1.5.1
[pip3] torchvision==0.15.2
[pip3] triton==2.0.0
[conda] Could not collect
```[tasklist]
### Tasks
```

cc @mruberry @kurtamohler @ezyang @msaroufim @wconstab @bdhirsh @anijain2305 @EikanWang @jgong5 @wenzhe-nrv @sanchitintel

**Comment by ChijinZ** [comment]
I found another testcase that triggers similar issue. The simplified code is:
```python
import numpy as np
import torch
import pickle

class M(torch.nn.Module):
    def __init__(self):
        super().__init__()
        self.x = torch.nn.Parameter(torch.tensor([6, 7, 4, 4, 5, 4, 6, 6, 7, 3, 4, 5, 3, 4, 5, 5, 7, 6, 4, 4, 7, 6, 6, 4, 4, 6, 7, 3, 7, 3, 6, 5, 6, 5, 5, 5, 5, 4, 6, 4, 7, 7, 4, 6, 5, 4, 4, 5, 4, 7, 7], dtype=torch.int16), requires_grad=False)
    def forward(self, *args):
        _args = args
        getitem = _args[0];  _args = None

        cat = torch.cat((self.x, getitem), dim = 0);  _tensor_constant0 = getitem = None
        sum_1 = cat.sum(0)
        return sum_1

m = M()

inp = [np.array([3], dtype='int16')]

opt = torch.compile(m, fullgraph=True, backend='inductor', mode=None)

m_out = m(*[torch.from_numpy(v).to('cpu') for v in inp]).cpu().detach()

opt_out = opt(*[torch.from_numpy(v).to('cpu') for v in inp]).cpu().detach()

print(m_out)
print(opt_out)
```

**Comment by jgong5** [comment]
I cannot reproduce the issue with the latest mainline: 41bd0fde7e886d7b08097e9a1123baa08e18c929. Can you have a quick try to see if the issue is still there? Thanks.

Collecting environment information...
PyTorch version: 2.2.0a0+git41bd0fd
Is debug build: False
CUDA used to build PyTorch: None
ROCM used to build PyTorch: N/A

OS: Ubuntu 22.04.1 LTS (x86_64)
GCC version: (Ubuntu 11.4.0-1ubuntu1~22.04) 11.4.0
Clang version: 14.0.0-1ubuntu1.1
CMake version: version 3.27.2
Libc version: glibc-2.35

Python version: 3.9.16 (main, Mar  8 2023, 14:00:05)  [GCC 11.2.0] (64-bit runtime)
Python platform: Linux-5.15.0-78-generic-x86_64-with-glibc2.35
Is CUDA available: False
CUDA runtime version: No CUDA
CUDA_MODULE_LOADING set to: N/A
GPU models and configuration: No CUDA
Nvidia driver version: No CUDA
cuDNN version: No CUDA
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

CPU:
Architecture:                    x86_64
CPU op-mode(s):                  32-bit, 64-bit
Address sizes:                   52 bits physical, 57 bits virtual
Byte Order:                      Little Endian
CPU(s):                          128
On-line CPU(s) list:             0-127
Vendor ID:                       GenuineIntel
Model name:                      Intel(R) Xeon(R) Platinum 8358 CPU @ 2.60GHz
CPU family:                      6
Model:                           106
Thread(s) per core:              2
Core(s) per socket:              32
Socket(s):                       2
Stepping:                        6
CPU max MHz:                     3400.0000
CPU min MHz:                     800.0000
BogoMIPS:                        5200.00
Flags:                           fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush dts acpi mmx fxsr sse sse2 ss ht tm pbe syscall nx pdpe1gb rdtscp lm constant_tsc art arch_perfmon pebs bts rep_good nopl xtopology nonstop_tsc cpuid aperfmperf pni pclmulqdq dtes64 monitor ds_cpl smx est tm2 ssse3 sdbg fma cx16 xtpr pdcm pcid dca sse4_1 sse4_2 x2apic movbe popcnt tsc_deadline_timer aes xsave avx f16c rdrand lahf_lm abm 3dnowprefetch cpuid_fault epb cat_l3 invpcid_single ssbd mba ibrs ibpb stibp ibrs_enhanced fsgsbase tsc_adjust bmi1 avx2 smep bmi2 erms invpcid cqm rdt_a avx512f avx512dq rdseed adx smap avx512ifma clflushopt clwb intel_pt avx512cd sha_ni avx512bw avx512vl xsaveopt xsavec xgetbv1 xsaves cqm_llc cqm_occup_llc cqm_mbm_total cqm_mbm_local split_lock_detect wbnoinvd dtherm ida arat pln pts hwp hwp_act_window hwp_epp hwp_pkg_req avx512vbmi umip pku ospke avx512_vbmi2 gfni vaes vpclmulqdq avx512_vnni avx512_bitalg tme avx512_vpopcntdq la57 rdpid fsrm md_clear pconfig flush_l1d arch_capabilities
L1d cache:                       3 MiB (64 instances)
L1i cache:                       2 MiB (64 instances)
L2 cache:                        80 MiB (64 instances)
L3 cache:                        96 MiB (2 instances)
NUMA node(s):                    2
NUMA node0 CPU(s):               0-31,64-95
NUMA node1 CPU(s):               32-63,96-127
Vulnerability Itlb multihit:     Not affected
Vulnerability L1tf:              Not affected
Vulnerability Mds:               Not affected
Vulnerability Meltdown:          Not affected
Vulnerability Mmio stale data:   Mitigation; Clear CPU buffers; SMT vulnerable
Vulnerability Retbleed:          Not affected
Vulnerability Spec store bypass: Mitigation; Speculative Store Bypass disabled via prctl and seccomp
Vulnerability Spectre v1:        Mitigation; usercopy/swapgs barriers and __user pointer sanitization
Vulnerability Spectre v2:        Mitigation; Enhanced IBRS, IBPB conditional, RSB filling, PBRSB-eIBRS SW sequence
Vulnerability Srbds:             Not affected
Vulnerability Tsx async abort:   Not affected

Versions of relevant libraries:
[pip3] intel-extension-for-pytorch==2.1.0+git5be28db
[pip3] numpy==1.24.3
[pip3] torch==2.2.0a0+git41bd0fd
[pip3] torchvision==0.15.2
[pip3] triton==2.0.0
[conda] intel-extension-for-pytorch 2.1.0+git5be28db           dev_0    <develop>
[conda] mkl                       2023.1.0                 pypi_0    pypi
[conda] mkl-include               2023.1.0                 pypi_0    pypi
[conda] mkl-static                2023.1.0                 pypi_0    pypi
[conda] numpy                     1.24.3                   pypi_0    pypi
[conda] torch                     2.2.0a0+git41bd0fd           dev_0    <develop>
[conda] torchvision               0.15.2                   pypi_0    pypi
[conda] triton                    2.0.0                    pypi_0    pypi

**Comment by ChijinZ** [comment]
I just used the latest stable version of torch (i.e., ``2.0.1+cu117``) using ``pip3 install torch torchvision torchaudio
``. I did not try to build torch from scatch...

Should I try the nightly pre-built version?

**Comment by jgong5** [comment]
> Should I try the nightly pre-built version?

Yes, can you try that out?

**Comment by ChijinZ** [comment]
> Yes, can you try that out?

I tried nightly pre-built version (2.2.0.dev20230911+cu118). The outputs are consistent. It seems that the bug was fixed between versions 2.0.1 and 2.2.0.

**Comment by jgong5** [comment]
> > Yes, can you try that out?
> 
> I tried nightly pre-built version (2.2.0.dev20230911+cu118). The outputs are consistent. It seems that the bug was fixed between versions 2.0.1 and 2.2.0.

Thanks for the confirmation. I will close this issue then.
